llvm/lib/Target/AMDGPU/SIFoldOperands.cpp

```diff
@@ -1147,10 +1218,12 @@ void SIFoldOperandsImpl::foldOperand(
   if (UseOp->isReg() && OpToFold.isReg()) {
     if (UseOp->isImplicit())
       return;
-    // Allow folding from SGPRs to 16-bit VGPRs.
+    // Allow folding from SGPRs to 16-bit VGPRs
+    // or any VGPR if the ultimate source is SGPR.
     if (UseOp->getSubReg() != AMDGPU::NoSubRegister &&
         (UseOp->getSubReg() != AMDGPU::lo16 ||
-         !TRI->isSGPRReg(*MRI, OpToFold.getReg())))
+         !TRI->isSGPRReg(*MRI, OpToFold.getReg())) &&
+        !checkIsSourceSGPR(OpToFold, *MRI, ST->getRegisterInfo()))
```

Comment:
If I change the entire condition to just be this instead:
```cpp
    if (UseOp->getSubReg() != AMDGPU::NoSubRegister &&
         !TRI->isSGPRReg(*MRI, OpToFold.getReg())) 
```
It appears to pass the lit tests in the same manner as this current change, without needing to look and see if the ultimate source is SGPR.

I am not certain if this check for lo16 is here for some reason that is not captured in the lit tests, so I have included what I have for the moment.